Writers failed with NoSuchSegmentException: Segment does not exist

While exercising longevity IO workload (4 Reader, 3 Writer, scale type as EVENTS_PER_SECOND with retention enabled) all writers have been failed with `NoSuchSegmentException: Segment does not exist` and `ERROR i.p.l.t.w.writers.PravegaWriter - Writer finished with Error`

Environment details: PKS / K8 with medium cluster:
```
3 master: xlarge: 4 CPU, 16 GB Ram, 32 GB Disk
5 worker: 2xlarge: 8 CPU, 32 GB Ram, 64 GB Disk
Tier-1 storage is from VSAN datastore
Tier-2 storage curved on NFS Client provisioner using Isilon as backend
```
```
Pravega version: 0.5.0-2260.b477eec
Zookeeper Operator : pravega/zookeeper-operator:0.2.1
Pravega Operator: pravega/pravega-operator:0.3.2 
```
All 3 writer failed after ~ 5hours of operation
```
INFO  [2019-05-26 00:42:41,885] io.pravega.longevity.utils.PerformanceUtils: Got new data  Running for 5h 35m
INFO  [2019-05-26 00:42:41,885] io.pravega.longevity.utils.PerformanceUtils: Writers (0/3): events:774,207, events/sec:38, KB/sec:29.486328
```
Following are exception on `PravegaWriter - Writer finished with Error`
```
2019-05-26 00:42:27,081 20122492 [ScalingRetransmition-longevity/smallScale-85] ERROR i.p.l.t.w.writers.PravegaWriter - Writer finished with Error
io.pravega.client.segment.impl.NoSuchSegmentException: Segment does not exist:longevity/smallScale/775.#epoch.486
        at io.pravega.client.stream.impl.SegmentSelector.lambda$refreshSegmentEventWritersUponSealed$1(SegmentSelector.java:95)
        at java.util.ArrayList.forEach(Unknown Source)
        at io.pravega.client.stream.impl.SegmentSelector.refreshSegmentEventWritersUponSealed(SegmentSelector.java:94)
        at io.pravega.client.stream.impl.EventStreamWriterImpl.lambda$null$1(EventStreamWriterImpl.java:159)
        at io.pravega.common.util.Retry$RetryAndThrowBase.run(Retry.java:173)
        at io.pravega.client.stream.impl.EventStreamWriterImpl.lambda$handleLogSealed$2(EventStreamWriterImpl.java:146)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.lang.Thread.run(Unknown Source)
```
```
2019-05-26 00:42:27,082 20122493 [grpc-default-executor-515] WARN  i.p.c.stream.impl.ControllerImpl - gRPC call for getSuccessors with trace id 0 failed with server error.
io.grpc.StatusRuntimeException: UNAVAILABLE: Keepalive failed. The connection is likely gone
        at io.grpc.Status.asRuntimeException(Status.java:530)
        at io.grpc.stub.ClientCalls$StreamObserverToCallListenerAdapter.onClose(ClientCalls.java:434)
        at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
        at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
        at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
        at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
        at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
        at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
        at io.grpc.internal.CensusStatsModule$StatsClientInterceptor$1$1.onClose(CensusStatsModule.java:694)
        at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
        at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
        at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
        at io.grpc.internal.CensusTracingModule$TracingClientInterceptor$1$1.onClose(CensusTracingModule.java:397)
        at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:459)
        at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:63)
        at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.close(ClientCallImpl.java:546)
        at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.access$600(ClientCallImpl.java:467)
        at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:584)
        at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
        at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.lang.Thread.run(Unknown Source)
```

My findings from the logs:

- a.  The writer is trying to fetch the successors for a given segment by contacting the controller.
- b.  The controller `getSuccessor` call failed. (the controller client has internal retries configured in case of an error and the retries to contact the controller failed too indicating that the controller is not reachable). The client logs this error ref [1]
- c. Incase of an error or an actual `null` returned by the controller (in case stream is deleted) the client 
propagates this error to all the segment writers and ensures `NoSuchSegmentException` exception is thrown. The writer cannot proceed further if the call to the controller fails. ref [2]

[1]
```
2019-05-26 00:42:27,078 20122489 [ScalingRetransmition-longevity/smallScale-85] ERROR i.p.c.stream.impl.SegmentSelector - Error while fetching successors for segment: longevity/smallScale/775.#epoch.
486
io.pravega.common.util.RetriesExhaustedException: java.util.concurrent.CompletionException: io.grpc.StatusRuntimeException: UNAVAILABLE: Keepalive failed. The connection is likely gone
        at io.pravega.common.util.Retry$RetryAndThrowBase.lambda$null$6(Retry.java:245)
        at java.util.concurrent.CompletableFuture.uniExceptionally(Unknown Source)
        at java.util.concurrent.CompletableFuture$UniExceptionally.tryFire(Unknown Source)
        at java.util.concurrent.CompletableFuture.postComplete(Unknown Source)
        at java.util.concurrent.CompletableFuture.completeExceptionally(Unknown Source)
        at io.pravega.client.stream.impl.ControllerImpl$RPCAsyncCallback.onError(ControllerImpl.java:1018)
        at io.grpc.stub.ClientCalls$StreamObserverToCallListenerAdapter.onClose(ClientCalls.java:434)
        at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
        at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
        at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
        at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
        at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
        at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
        at io.grpc.internal.CensusStatsModule$StatsClientInterceptor$1$1.onClose(CensusStatsModule.java:694)
        at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
        at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
        at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
        at io.grpc.internal.CensusTracingModule$TracingClientInterceptor$1$1.onClose(CensusTracingModule.java:397)
        at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:459)
        at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:63)
        at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.close(ClientCallImpl.java:546)
        at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.access$600(ClientCallImpl.java:467)
        at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:584)
        at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
        at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.lang.Thread.run(Unknown Source)
Caused by: java.util.concurrent.CompletionException: io.grpc.StatusRuntimeException: UNAVAILABLE: Keepalive failed. The connection is likely gone
        at java.util.concurrent.CompletableFuture.encodeRelay(Unknown Source)
        at java.util.concurrent.CompletableFuture.completeRelay(Unknown Source)
        at java.util.concurrent.CompletableFuture.uniRelay(Unknown Source)
        at java.util.concurrent.CompletableFuture$UniRelay.tryFire(Unknown Source)
        ... 25 common frames omitted
Caused by: io.grpc.StatusRuntimeException: UNAVAILABLE: Keepalive failed. The connection is likely gone
        at io.grpc.Status.asRuntimeException(Status.java:530)
        ... 22 common frames omitted
```
[2]
```
2019-05-26 00:42:27,081 20122492 [ScalingRetransmition-longevity/smallScale-85] ERROR i.p.l.t.w.writers.PravegaWriter - Writer finished with Error
io.pravega.client.segment.impl.NoSuchSegmentException: Segment does not exist:longevity/smallScale/775.#epoch.486
        at io.pravega.client.stream.impl.SegmentSelector.lambda$refreshSegmentEventWritersUponSealed$1(SegmentSelector.java:95)
        at java.util.ArrayList.forEach(Unknown Source)
        at io.pravega.client.stream.impl.SegmentSelector.refreshSegmentEventWritersUponSealed(SegmentSelector.java:94)
        at io.pravega.client.stream.impl.EventStreamWriterImpl.lambda$null$1(EventStreamWriterImpl.java:159)
        at io.pravega.common.util.Retry$RetryAndThrowBase.run(Retry.java:173)
        at io.pravega.client.stream.impl.EventStreamWriterImpl.lambda$handleLogSealed$2(EventStreamWriterImpl.java:146)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.lang.Thread.run(Unknown Source)
```

--- a/client/src/main/java/io/pravega/client/stream/impl/SegmentSelector.java
+++ b/client/src/main/java/io/pravega/client/stream/impl/SegmentSelector.java
@@ -18,6 +18,7 @@ import io.pravega.client.stream.EventWriterConfig;
 import io.pravega.client.stream.Stream;
 import io.pravega.common.concurrent.Futures;
 import io.pravega.common.hash.RandomFactory;
+import io.pravega.common.util.RetriesExhaustedException;
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.HashMap;
@@ -80,19 +81,17 @@ public class SegmentSelector {
         return currentSegments.getSegmentForKey(routingKey);
     }
 
-    public List<PendingEvent> refreshSegmentEventWritersUponSealed(Segment sealedSegment, Consumer<Segment>
-            segmentSealedCallback) {
+    public List<PendingEvent> refreshSegmentEventWritersUponSealed(Segment sealedSegment, Consumer<Segment> segmentSealedCallback) {
         StreamSegmentsWithPredecessors successors = Futures.getAndHandleExceptions(
                 controller.getSuccessors(sealedSegment), t -> {
                     log.error("Error while fetching successors for segment: {}", sealedSegment, t);
+                    // Remove all writers and fail all pending writes
+                    Exception e = (t instanceof RetriesExhaustedException) ? new ControllerFailureException(t) : new NoSuchSegmentException(sealedSegment.toString());
+                    removeAllWriters().forEach(event -> event.getAckFuture().completeExceptionally(e));
                     return null;
                 });
 
         if (successors == null) {
-            // Stream is deleted, complete all pending writes exceptionally.
-            log.error("Stream is deleted, all pending writes will be completed exceptionally");
-            removeAllWriters().forEach(event -> event.getAckFuture()
-                    .completeExceptionally(new NoSuchSegmentException(sealedSegment.toString())));
             return Collections.emptyList();
         } else {
             return updateSegmentsUponSealed(successors, sealedSegment, segmentSealedCallback);

--- a/client/src/test/java/io/pravega/client/stream/impl/SegmentSelectorTest.java
+++ b/client/src/test/java/io/pravega/client/stream/impl/SegmentSelectorTest.java
@@ -10,12 +10,15 @@
 package io.pravega.client.stream.impl;
 
 import com.google.common.collect.ImmutableList;
+import io.grpc.Status;
+import io.grpc.StatusRuntimeException;
 import io.pravega.client.segment.impl.NoSuchSegmentException;
 import io.pravega.client.segment.impl.Segment;
 import io.pravega.client.segment.impl.SegmentOutputStream;
 import io.pravega.client.segment.impl.SegmentOutputStreamFactory;
 import io.pravega.client.stream.EventWriterConfig;
 
+import io.pravega.common.util.RetriesExhaustedException;
 import java.nio.ByteBuffer;
 import java.util.Arrays;
 import java.util.Collections;
@@ -167,6 +170,51 @@ public class SegmentSelectorTest {
 
         assertEquals(Collections.emptyList(), selector.refreshSegmentEventWritersUponSealed(segment0, segmentSealedCallback));
         assertFutureThrows("Writer Future", writerFuture, t -> t instanceof NoSuchSegmentException);
+    }
+
+    @Test
+    public void testControllerNotReachable() {
+        final Segment segment0 = new Segment(scope, streamName, 0);
+        final Segment segment1 = new Segment(scope, streamName, 1);
+        final CompletableFuture<Void> writerFuture = new CompletableFuture<>();
+
+        // Setup Mock.
+        SegmentOutputStream s0Writer = Mockito.mock(SegmentOutputStream.class);
+        SegmentOutputStream s1Writer = Mockito.mock(SegmentOutputStream.class);
+        when(s0Writer.getUnackedEventsOnSeal())
+                .thenReturn(ImmutableList.of(PendingEvent.withHeader("0", ByteBuffer.wrap("e".getBytes()), writerFuture)));
+
+        SegmentOutputStreamFactory factory = Mockito.mock(SegmentOutputStreamFactory.class);
+        when(factory.createOutputStreamForSegment(eq(segment0), ArgumentMatchers.<Consumer<Segment>>any(), any(EventWriterConfig.class), anyString()))
+                .thenReturn(s0Writer);
+        when(factory.createOutputStreamForSegment(eq(segment1), ArgumentMatchers.<Consumer<Segment>>any(), any(EventWriterConfig.class), anyString()))
+                .thenReturn(s1Writer);
+
+        Controller controller = Mockito.mock(Controller.class);
+        SegmentSelector selector = new SegmentSelector(new StreamImpl(scope, streamName), controller, factory, config);
+        TreeMap<Double, SegmentWithRange> segments = new TreeMap<>();
+        addNewSegment(segments, 0, 0.0, 0.5);
+        addNewSegment(segments, 1, 0.5, 1.0);
+        StreamSegments streamSegments = new StreamSegments(segments, "");
+
+        when(controller.getCurrentSegments(scope, streamName))
+                .thenReturn(CompletableFuture.completedFuture(streamSegments));
+        //trigger refresh.
+        selector.refreshSegmentEventWriters(segmentSealedCallback);
+
+        //simulate controller failure when controller.getSuccessors() is invoked.
+        when(controller.getSuccessors(segment0))
+                .thenAnswer(i -> {
+                    CompletableFuture<StreamSegmentsWithPredecessors> result = new CompletableFuture<>();
+                    // Controller client in case of RPC exceptions throws a StatusRuntimeException which is retried by the client.
+                    // If the controller client is not able to reach the controller after all the retries a RetriesExhaustedException is
+                    // thrown.
+                    result.completeExceptionally(new RetriesExhaustedException(new StatusRuntimeException(Status.DATA_LOSS)));
+                    return result;
+                });
+
+        assertEquals(Collections.emptyList(), selector.refreshSegmentEventWritersUponSealed(segment0, segmentSealedCallback));
+        assertFutureThrows("Writer Future", writerFuture, t -> t instanceof ControllerFailureException);
     }
 
 }

